[Scala] added new scala-cask generator for the cask framework

aaronp opened the pull request:
This change adds a new scala-cask generator for the [cask](https://github.com/com-lihaoyi/cask) framework.

Thanks in advance to scala committee reviewers @clasnake, @jimschubert, @shijinkui, @ramzimaalej, @chameleon82, @Bouillie and/or @fish86

<!-- Please check the completed items below -->
### PR checklist
 
- [ x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.1.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

wing328 commented on .gitignore:
```diff
@@ -282,3 +282,4 @@ samples/openapi3/client/petstore/go/privatekey.pem
 
 ## OCaml
 samples/client/petstore/ocaml/_build/
+samples
```
Thanks for the PR. Why add this line?

aaronp replied:
Ah - I think that's a hang over from what I was using as an output directory. I'll remove it. Thanks

wing328 commented on .run/CaskCodeGen.run.xml:
```diff
@@ -0,0 +1,25 @@
+      <option name="Make" enabled="true" />
+    </method>
+  </configuration>
+</component>
```
what's the purpose of this file?

aaronp replied:
It was a run profile for testing -- something I can just take out

wing328 commented:
thanks for the PR.

can you please also add the new sample folder to .github/workflows/samples-scala.yaml so that the CI can test it moving forward?

wing328 commented:
please ping me via Slack if you need help with this PR.

https://join.slack.com/t/openapi-generator/shared_invite/zt-12jxxd7p2-XUeQM~4pzsU9x~eGLQqX2g

thanks again for the PR

aaronp commented:
I think I've addressed the comments in the PR - many thanks @wing328 for the review!

aaronp commented:
How very exciting! Thanks all for your time and patience in reviewing -- very much appreciated!

aaronp commented:
Thanks @wing328 for the review and help. I've just updated the branch and included server validation improvements, so hopefully is ready for another review?

wing328 commented:
can you please take a look at the CI build failure when you've time?

https://github.com/OpenAPITools/openapi-generator/actions/runs/8440910756/job/23134572125?pr=18046

let me know if you need help to fix it.

wing328 commented:
the scala cask tests also failed: https://github.com/OpenAPITools/openapi-generator/actions/runs/8440910764/job/23134573837?pr=18046

please take a look when you've time

wing328 commented:
Thanks for the PR but your commit (as shown in the Commits tab) is not linked to your Github account, which means this PR won't count as your contribution in https://github.com/OpenAPITools/openapi-generator/graphs/contributors.

Let me know if you need help fixing it.

Ref: https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-can-i-update-commits-that-are-not-linked-to-my-github-account

wing328 commented:
the test failed. can you please take a look?

aaronp commented:
Thanks William -
Looks like I’ve got some work to do!

Thanks so much for the helpful comments 👍

On Sat, 30 Mar 2024 at 13:27, William Cheng ***@***.***>
wrote:

> the test failed. can you please take a look?
>
>
> https://github.com/OpenAPITools/openapi-generator/actions/runs/8468026875/job/23262734520?pr=18046
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenAPITools/openapi-generator/pull/18046#issuecomment-2028058536>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABMIOKYFZNRMO3SLDDLMU3Y224VZAVCNFSM6AAAAABEJ32LSOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMRYGA2TQNJTGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

aaronp commented:
So, for the first failure
<https://github.com/OpenAPITools/openapi-generator/actions/runs/8440910756/job/23134572125?pr=18046>,
I have no idea what this is / where it comes from / if it's from my PR:


java.lang.IllegalArgumentException: Null key for a Map not allowed in JSON
(use a converting NullKeySerializer?) (through reference chain:
java.util.LinkedHashMap["components"]->java.util.LinkedHashMap["schemas"]->java.util.LinkedHashMap["Good"]->java.util.LinkedHashMap["properties"]->java.util.LinkedHashMap["null"])



It seems to be innocuous / doesn't affect things now, though I can
reproduce that / see that failure locally.

I've added my ***@***.***' as a secondary email to my Github
account -- thanks for that!

It looks like a lot of these failures
<https://github.com/OpenAPITools/openapi-generator/actions/runs/8468026875/job/23262734520?pr=18046>
are for other scala projects (lagom, scalaz, finch). I'll double-check, but
I can't see how my PR is affecting these other projects?
[image: image.png]


For the cask one I've introduced
<https://github.com/OpenAPITools/openapi-generator/actions/runs/8468026875/job/23262735057?pr=18046#step:5:40>,
I can't reproduce that locally. I've run the steps:

./mvnw clean package
 ./bin/generate-samples.sh ./bin/configs/*.yaml

co
At this point I've checked the code generated
in /samples/server/petstore/scala/cask and it all builds ok for me /
compiles and tests just fine?
I should note, the compile step does reformat the code, so I'm not sure if
that has an impact/makes a difference?

I regularly sink my fork and rebuild.

I'll double-check my changes if they can affect the other scala projects,
but not quite sure where else to go from here.

As ever, thanks so much for your time and help!

Best,
Aaron



On Sat, Mar 30, 2024 at 1:36 PM Aaron Pritzlaff ***@***.***>
wrote:

> Thanks William -
> Looks like I’ve got some work to do!
>
> Thanks so much for the helpful comments 👍
>
> On Sat, 30 Mar 2024 at 13:27, William Cheng ***@***.***>
> wrote:
>
>> the test failed. can you please take a look?
>>
>>
>> https://github.com/OpenAPITools/openapi-generator/actions/runs/8468026875/job/23262734520?pr=18046
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/OpenAPITools/openapi-generator/pull/18046#issuecomment-2028058536>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AABMIOKYFZNRMO3SLDDLMU3Y224VZAVCNFSM6AAAAABEJ32LSOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMRYGA2TQNJTGY>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

wing328 commented:
can you review and fix: https://github.com/OpenAPITools/openapi-generator/actions/runs/8492216876/job/23265283400?pr=18046?

for test failures with other scala samles, please ignore those for the time being

aaronp commented on samples/server/petstore/scala/cask/src/main/scala/org/openapitools/server/api/PetRoutes.scala:
```diff
@@ -0,0 +1,229 @@
+// this is generated from apiRoutes.mustache
+package org.openapitools.server.api
+
+import org.openapitools.server.model.*
```
This wildcard import should include the PetData class

wing328 commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaCaskServerCodegen.java:
```diff
@@ -0,0 +1,957 @@
+        public OperationGroup(String httpMethod, String pathPrefix) {
+            this.httpMethod = httpMethod;
+            this.pathPrefix = pathPrefix;
+            caskAnnotation = "@cask." + httpMethod.toLowerCase();
```
for toLowerCase, you will need to use Locale.ROOT to address the build failure:

https://github.com/OpenAPITools/openapi-generator/blob/1704ae1194c7e7b07f7522a5ab038bf002c0f5d4/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AdaCodegen.java#L125

aaronp replied:
Boom! Smashing through these. Fixed - thanks

wing328 replied:
can you please also fix the rest?

```
Forbidden method invocation: java.lang.String#toUpperCase() [Uses default locale]
Error:    in org.openapitools.codegen.languages.ScalaCaskServerCodegen (ScalaCaskServerCodegen.java:376)
Error:  Forbidden method invocation: java.lang.String#toUpperCase() [Uses default locale]
Error:    in org.openapitools.codegen.languages.ScalaCaskServerCodegen (ScalaCaskServerCodegen.java:37[9](https://github.com/OpenAPITools/openapi-generator/actions/runs/8505836811/job/23295001534?pr=18046#step:5:10))
Error:  Scanned 529 class file(s) for forbidden API invocations (in 1.00s), 2 error(s).
Error:  Failed to execute goal de.thetaphi:forbiddenapis:3.5.1:check (default) on project openapi-generator: Check for forbidden API calls failed, see log. -> [Help 1]
```

aaronp replied:
yepper - done

wing328 commented:
can you please take a look at https://github.com/OpenAPITools/openapi-generator/actions/runs/8505982227/job/23295571166?pr=18046 ?

wing328 commented:
samples seems not up-to-date: https://github.com/OpenAPITools/openapi-generator/actions/runs/8505982218/job/23295650385?pr=18046

please regenerate the samples

wing328 commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaCaskServerCodegen.java:
```diff
@@ -0,0 +1,965 @@
+                array = pathPrefix.split(File.separator, -1);
+            } catch (PatternSyntaxException exp) {
+                throw new RuntimeException("For some reason, splitting >" + pathPrefix + "< on >" + File.separator + "< threw " + exp, exp);
+            }
```
looks like this line is throwing exception in windows?

what about using just `/` instead of File.separator, which is `\` in windows?

aaronp replied:
good catch - done

wing328 commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaCaskServerCodegen.java:
```diff
@@ -0,0 +1,958 @@
+    protected String gitUserId = env("GIT_USER", System.getProperty("user.name"));
+    protected String gitRepoId = env("GIT_REPO", artifactId);
+    protected String appName = env("APP_NAME", "Cask App");
+    protected String infoEmail = env("INFO_EMAIL", "contact@kindservices.co.uk");
```
for gitUserId and gitRepoId,  you can use these directly from default codegen.

please do not use environment variables, and use additional properties instead. e.g.  https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaAkkaHttpServerCodegen.java#L174-L224


infoEmail should be obtained from the spec so I don't think we need to set it here

aaronp replied:
Many thanks.

I've reworked these properties and cleaned up a lot of the debug code.

I've changed the sample properties and output destination -- I hope that's not too confusing to follow from the original PR

aaronp commented:
@wing328, it looks like this might be good to go, unless I've missed something?

wing328 commented on .github/workflows/samples-scala.yaml:
```diff
@@ -32,6 +32,7 @@ jobs:
           - samples/server/petstore/scalatra
           - samples/server/petstore/scala-finch # cannot be tested with jdk11
           - samples/server/petstore/scala-http4s-server
+          - samples/server/petstore/scala/cask
```
should it be `samples/server/petstore/scala-cask` instead?

aaronp replied:
yes - good catch. updated

wing328 commented on samples/server/petstore/scala-cask/src/main/scala/mo/del/pack/age/ApiResponse.scala:
```diff
@@ -0,0 +1,53 @@
+/**
```
the  "mo/del" folder looks a bit weird to me.
should it be simply just "model"?

aaronp replied:
I was just testing out weird/nested configs, but you're right. I've updated it now to be a more standard package structure

aaronp commented:
Thanks again @wing328 -- I've removed the sample client code (I'm not entirely sure how that even got there) and the .bsp sample directory as that was generated from me then compiling the sample code.

wing328 commented:
i've filed https://github.com/OpenAPITools/openapi-generator/pull/18344 based on this PR with some minor enhancements.

wing328 commented:
merged https://github.com/OpenAPITools/openapi-generator/pull/18344 (commits still authored by you).

thanks for the new generator 👍
